File: backend/shared/src/domain/entities/ehr/visit.rs
//! EHR Visit/Encounter Entity
//!
//! Corresponds to VistA File #9000010 (^AUPNVSIT) - Visit/Encounter

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;
use uuid::Uuid;

use crate::shared::AuditFields;

/// Visit type classification
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "VARCHAR")]
#[sqlx(rename_all = "lowercase")]
pub enum VisitType {
    #[serde(rename = "outpatient")]
    Outpatient,
    #[serde(rename = "inpatient")]
    Inpatient,
    #[serde(rename = "emergency")]
    Emergency,
    #[serde(rename = "telehealth")]
    Telehealth,
    #[serde(rename = "home")]
    Home,
    #[serde(rename = "observation")]
    Observation,
}

impl Default for VisitType {
    fn default() -> Self {
        VisitType::Outpatient
    }
}

/// Visit status
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "VARCHAR")]
#[sqlx(rename_all = "lowercase")]
pub enum VisitStatus {
    #[serde(rename = "scheduled")]
    Scheduled,
    #[serde(rename = "checked_in")]
    CheckedIn,
    #[serde(rename = "in_progress")]
    InProgress,
    #[serde(rename = "completed")]
    Completed,
    #[serde(rename = "cancelled")]
    Cancelled,
    #[serde(rename = "no_show")]
    NoShow,
}

impl Default for VisitStatus {
    fn default() -> Self {
        VisitStatus::Scheduled
    }
}

/// EHR Visit Entity
///
/// Represents a patient encounter/visit in the EHR system.
/// Corresponds to VistA File #9000010 (^AUPNVSIT - Visit).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EhrVisit {
    pub id: Uuid,

    /// Internal Entry Number (VistA IEN)
    pub ien: i64,

    /// Organization (multi-tenant)
    pub organization_id: Uuid,

    /// Patient ID (FK to ehr_patients)
    pub patient_id: Uuid,

    /// Visit type (outpatient, inpatient, etc.)
    pub visit_type: VisitType,

    /// Visit status
    pub status: VisitStatus,

    /// Visit date/time
    pub visit_datetime: DateTime<Utc>,

    /// Check-in time
    pub check_in_time: Option<DateTime<Utc>>,

    /// Check-out time
    pub check_out_time: Option<DateTime<Utc>>,

    /// Location (clinic, hospital, etc.)
    pub location_id: Option<Uuid>,
    pub location_name: Option<String>,

    /// Primary provider
    pub provider_id: Option<Uuid>,
    pub provider_name: Option<String>,

    /// Chief complaint
    pub chief_complaint: Option<String>,

    /// Reason for visit (CPT E/M level or custom)
    pub reason_for_visit: Option<String>,

    /// Service category (medicine, surgery, etc.)
    pub service_category: Option<String>,

    /// Disposition (discharged, admitted, transferred)
    pub disposition: Option<String>,

    /// MUMPS global data (for flexible storage)
    pub mumps_data: Option<JsonValue>,

    // Audit fields
    pub request_id: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub created_by: Option<Uuid>,
    pub updated_by: Option<Uuid>,
    pub system_id: Option<String>,
    pub version: i64,
}

impl EhrVisit {
    /// Create a new visit
    pub fn new(
        organization_id: Uuid,
        patient_id: Uuid,
        visit_type: VisitType,
        visit_datetime: DateTime<Utc>,
    ) -> Self {
        let audit = AuditFields::new();
        Self {
            id: Uuid::new_v4(),
            ien: 0,
            organization_id,
            patient_id,
            visit_type,
            status: VisitStatus::Scheduled,
            visit_datetime,
            check_in_time: None,
            check_out_time: None,
            location_id: None,
            location_name: None,
            provider_id: None,
            provider_name: None,
            chief_complaint: None,
            reason_for_visit: None,
            service_category: None,
            disposition: None,
            mumps_data: None,
            request_id: audit.request_id,
            created_at: audit.created_at,
            updated_at: audit.updated_at,
            created_by: audit.created_by,
            updated_by: audit.updated_by,
            system_id: audit.system_id,
            version: audit.version,
        }
    }

    /// Check in the patient
    pub fn check_in(&mut self) {
        self.status = VisitStatus::CheckedIn;
        self.check_in_time = Some(Utc::now());
        self.updated_at = Utc::now();
    }

    /// Start the visit
    pub fn start(&mut self) {
        self.status = VisitStatus::InProgress;
        self.updated_at = Utc::now();
    }

    /// Complete the visit
    pub fn complete(&mut self, disposition: Option<String>) {
        self.status = VisitStatus::Completed;
        self.check_out_time = Some(Utc::now());
        self.disposition = disposition;
        self.updated_at = Utc::now();
    }

    /// Cancel the visit
    pub fn cancel(&mut self) {
        self.status = VisitStatus::Cancelled;
        self.updated_at = Utc::now();
    }

    /// Mark as no-show
    pub fn mark_no_show(&mut self) {
        self.status = VisitStatus::NoShow;
        self.updated_at = Utc::now();
    }

    /// Check if visit is active (can be modified)
    pub fn is_active(&self) -> bool {
        matches!(
            self.status,
            VisitStatus::Scheduled | VisitStatus::CheckedIn | VisitStatus::InProgress
        )
    }

    /// Get visit duration in minutes (if completed)
    pub fn duration_minutes(&self) -> Option<i64> {
        match (self.check_in_time, self.check_out_time) {
            (Some(start), Some(end)) => {
                Some((end - start).num_minutes())
            }
            _ => None,
        }
    }
}

File: cli/packages/libs/shared/src/schemas/ehr/visit.ts
/**
 * EHR Visit/Encounter Schemas
 *
 * Runtime validation for patient visits and encounters.
 * Corresponds to VistA File #9000010 (^AUPNVSIT)
 * Single source of truth - use z.infer<> for types.
 */

import { z } from "zod";
import { DateTimeSchema } from "../common";
import { createTypeGuard, createAssertion } from "../guards";

// ============================================================================
// Visit Enums
// ============================================================================

/**
 * Visit type schema
 */
export const EhrVisitTypeSchema = z.enum([
  "outpatient",
  "inpatient",
  "emergency",
  "telehealth",
  "home",
  "observation",
]);

export type EhrVisitType = z.infer<typeof EhrVisitTypeSchema>;

/**
 * Visit status schema
 */
export const EhrVisitStatusSchema = z.enum([
  "scheduled",
  "checked_in",
  "in_progress",
  "completed",
  "cancelled",
  "no_show",
]);

export type EhrVisitStatus = z.infer<typeof EhrVisitStatusSchema>;

// ============================================================================
// Visit Schema
// ============================================================================

/**
 * Full EHR Visit schema
 */
export const EhrVisitSchema = z.object({
  // Identity
  id: z.string().uuid(),
  ien: z.number().int().positive(),
  organizationId: z.string().uuid(),
  patientId: z.string().uuid(),

  // Visit details
  visitType: EhrVisitTypeSchema,
  status: EhrVisitStatusSchema,
  visitDatetime: DateTimeSchema,
  checkInTime: DateTimeSchema.optional(),
  checkOutTime: DateTimeSchema.optional(),

  // Location and provider
  locationId: z.string().uuid().optional(),
  locationName: z.string().optional(),
  providerId: z.string().uuid().optional(),
  providerName: z.string().optional(),

  // Clinical details
  chiefComplaint: z.string().max(500).optional(),
  reasonForVisit: z.string().max(500).optional(),
  serviceCategory: z.string().max(200).optional(),
  disposition: z.string().max(200).optional(),

  // Audit fields
  createdAt: DateTimeSchema.optional(),
  createdBy: z.string().optional(),
  updatedAt: DateTimeSchema.optional(),
  updatedBy: z.string().optional(),

  // MUMPS data
  mumpsData: z.record(z.unknown()).optional(),
}).refine(
  // Business rule: If checked out, must have check-in time
  data => {
    if (data.checkOutTime) {
      return !!data.checkInTime;
    }
    return true;
  },
  {
    message: "Check-out time requires check-in time",
    path: ["checkInTime"],
  }
).refine(
  // Business rule: Check-out must be after check-in
  data => {
    if (data.checkOutTime && data.checkInTime) {
      return new Date(data.checkOutTime) > new Date(data.checkInTime);
    }
    return true;
  },
  {
    message: "Check-out time must be after check-in time",
    path: ["checkOutTime"],
  }
);

export type EhrVisit = z.infer<typeof EhrVisitSchema>;

// Type guards
export const isEhrVisit = createTypeGuard(EhrVisitSchema);

// Assertions
export const assertEhrVisit = createAssertion(EhrVisitSchema, 'EhrVisit');

// ============================================================================
// Create Visit Request Schema
// ============================================================================

export const CreateEhrVisitRequestSchema = z.object({
  patientId: z.string().uuid(),
  visitType: EhrVisitTypeSchema,
  visitDatetime: DateTimeSchema,
  locationId: z.string().uuid().optional(),
  providerId: z.string().uuid().optional(),
  chiefComplaint: z.string().max(500).optional(),
  reasonForVisit: z.string().max(500).optional(),
});

export type CreateEhrVisitRequest = z.infer<typeof CreateEhrVisitRequestSchema>;

export const isCreateEhrVisitRequest = createTypeGuard(CreateEhrVisitRequestSchema);
export const assertCreateEhrVisitRequest = createAssertion(CreateEhrVisitRequestSchema, 'CreateEhrVisitRequest');

// ============================================================================
// Update Visit Request Schema
// ============================================================================

export const UpdateEhrVisitRequestSchema = CreateEhrVisitRequestSchema.partial().extend({
  id: z.string().uuid(),
  status: EhrVisitStatusSchema.optional(),
  disposition: z.string().max(200).optional(),
});

export type UpdateEhrVisitRequest = z.infer<typeof UpdateEhrVisitRequestSchema>;

export const isUpdateEhrVisitRequest = createTypeGuard(UpdateEhrVisitRequestSchema);
export const assertUpdateEhrVisitRequest = createAssertion(UpdateEhrVisitRequestSchema, 'UpdateEhrVisitRequest');

// ============================================================================
// Visit Search Criteria Schema
// ============================================================================

export const EhrVisitSearchCriteriaSchema = z.object({
  patientId: z.string().uuid().optional(),
  visitType: EhrVisitTypeSchema.optional(),
  status: EhrVisitStatusSchema.optional(),
  providerId: z.string().uuid().optional(),
  locationId: z.string().uuid().optional(),
  dateFrom: DateTimeSchema.optional(),
  dateTo: DateTimeSchema.optional(),
});

export type EhrVisitSearchCriteria = z.infer<typeof EhrVisitSearchCriteriaSchema>;

export const isEhrVisitSearchCriteria = createTypeGuard(EhrVisitSearchCriteriaSchema);

// ============================================================================
// Utility Functions
// ============================================================================

export function isVisitCompleted(visit: EhrVisit): boolean {
  return visit.status === "completed";
}

export function isVisitInProgress(visit: EhrVisit): boolean {
  return visit.status === "in_progress";
}

export function getVisitDuration(visit: EhrVisit): number | null {
  if (!visit.checkInTime || !visit.checkOutTime) {
    return null;
  }

  const checkIn = new Date(visit.checkInTime);
  const checkOut = new Date(visit.checkOutTime);
  const durationMinutes = Math.floor((checkOut.getTime() - checkIn.getTime()) / (1000 * 60));

  return durationMinutes;
}

export function canCheckIn(visit: EhrVisit): boolean {
  return visit.status === "scheduled";
}

export function canCheckOut(visit: EhrVisit): boolean {
  return ["checked_in", "in_progress"].includes(visit.status);
}
